pkg/helm/helm.go

```diff
@@ -104,14 +104,79 @@ func (h *Hydrator) templateArgs(ctx context.Context, destDir string) ([]string,
 }
 
 func (h *Hydrator) appendValuesArgs(args []string) ([]string, error) {
-	valuesPath := filepath.Join(os.TempDir(), valuesFile)
-	if err := os.WriteFile(valuesPath, []byte(h.Values), 0644); err != nil {
-		return nil, fmt.Errorf("failed to create values file: %w", err)
+	var err error
+	switch h.ValuesMergeMode {
+
+	case "", "override":
+		for i, vf := range h.ValuesFrom {
+			if vf == "" {
+				continue
+			}
+			val, err := readFile(vf)
+			if err != nil {
+				return nil, err
+			}
+
+			klog.Infof("values from ConfigMap %s\n: %s\n", vf, val)
+			args, err = writeValuesPath([]byte(val), fmt.Sprintf("values-file-%d-", i), args)
+			if err != nil {
+				return nil, err
+			}
+		}
+		if len(h.Values) != 0 {
+			klog.Infof("inline values %s\n", h.Values)
+			args, err = writeValuesPath([]byte(h.Values), "", args)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+	case "merge":
```

Comment:
These modes should be defined as constants somewhere

Also I find the name `merge` a bit misleading here, as the other mode is still technically doing a merge but with a different strategy. Can we come up with a more precise name for this mode?

Reply:
Hmm... perhaps the ideal name would give the connotation that the files are being combined into one file before being passed to helm. Something like `combine`? Or `mergeFiles`? Do either of those sound more intuitive?

Reply:
The difference is how it treats the same key found in different values files. `override` overrides the values of the key with the one comes latter, whereas `merge` merges the values of the same key from different values files.

Reply:
We could update the CRD doc (i.e. comment in type) to make the behavior more clear (in relation to how keys are handled which users would understand better vs. how we implement it).

Reply:
Documenting the field in CRD helps, but I don't think it's sufficient. IMO the name itself should more clearly distinguish the difference between the modes.

As an example, I got confused when reading the e2e test case because the test behaves like:
- set up a `RootSync` with `valuesMergeMode` unset and with two values files
- verify that the two values files get merged
- set `valuesMergeMode=merge`
- verify that the two values files get merged, but this time merged with a different strategy

With a name like `valuesMergeMode=merge`, I would expect the alternative is that the values files don't get merged at all - for example the reconciler will stall/error if you define multiple values files.

Reply:
I think part of your confusion with the e2e was my fault, for not clarifying in the test that the default for `valuesMergeMode` is `override` (and I can add that). 

I do think the name `valuesMergeMode` already implies that the values are going to be merged in some way - and it wouldn't make sense for a user to put multiple values files in their RSync unless they were expecting them to be merged. I think `override` will be used in the vast majority of cases because it is equivalent to using multiple values files with helm CLI . 

I'm OK to rename `merge` to something else though, but I don't know if I can think of a great alternative. So unless we have other name candidates, and document properly what each option means and what the default is.

Reply:
> Sam, what I'm trying to say is that the we're merging vs. overriding values of the same keys in different values files, so the mode types makes sense to me.

The point I'm trying to make is that `override` (the default behavior) as defined here is just another strategy for merging two objects together in the conventional sense. When that's being contrasted against the other mode, `merge`, I see a user scratching their head wondering what `valuesMergeMode: merge` means in this context since `override` was also performing a different type of conventional merge. Perhaps from a different angle, merely observing that the value `merge` is a substring of the field `valuesMergeMode` seems like a bit of a smell to me. The alternative of `concatenate` that Natasha mentioned seems closer to a intuitive name to me, although she mentioned that isn't precise either.

I'm fine with conceding on this point if I'm getting overruled here

Reply:
> These modes should be defined as constants somewhere

Done.

I also updated the godoc/crd doc and added a comment to the test to make it more clear what the behavior is.

Reply:
As discussed offline, we will live it as `override` and `merge`, but rename the field to `valuesKeyMergeMode`.